sematic/abstract_future.py

```diff
@@ -81,9 +83,51 @@ class FutureProperties:
     inline: bool
     name: str
     tags: List[str]
+    kwargs: Dict[str, Any]
     resource_requirements: Optional[ResourceRequirements] = None
     retry_settings: Optional[RetrySettings] = None
     base_image_tag: Optional[str] = None
+    external_resources: List[ExternalResource] = field(default_factory=list)
+
+    def get_resolved_kwargs(self) -> Dict[str, Any]:
+        """
+        Extract only resolved/concrete kwargs.
+
+        For external resources, only ones in the ACTIVE state can be used
+        and are considered resolved.
+        """
+        resources_by_id = {r.id: r for r in self.external_resources}
+        resolved_kwargs = {}
+        for name, value in self.kwargs.items():
+            if isinstance(value, AbstractFuture):
+                if value.state == FutureState.RESOLVED:
+                    resolved_kwargs[name] = value.value
+            elif isinstance(value, ExternalResource):
+                value = resources_by_id.get(value.id)
+                if value.status.state == ResourceState.ACTIVE:
+                    resolved_kwargs[name] = value
```

Comment:
Surely you must have a mypy error here. `value` can be `None`. Maybe simply add a `not None` check to the `if` statement.